external-crates/move/move-vm/runtime/src/interpreter.rs

```diff
@@ -97,11 +99,128 @@ impl<'a, 'b> TypeView for TypeWithLoader<'a, 'b> {
     }
 }
 
+pub trait InterpreterInterface {
+    fn get_values_len(&self) -> usize;
+    fn set_location(&self, err: PartialVMError) -> VMError;
+    fn maybe_core_dump(&self, err: VMError, current_frame: &Frame) -> VMError;
+    fn get_internal_state(&self) -> ExecutionState;
+}
+
+impl InterpreterInterface for Interpreter {
+    fn get_values_len(&self) -> usize {
+        self.get_values_len()
+    }
+
+    fn set_location(&self, err: PartialVMError) -> VMError {
+        self.set_location(err)
+    }
+
+    fn maybe_core_dump(&self, err: VMError, current_frame: &Frame) -> VMError {
+        self.maybe_core_dump(err, current_frame)
+    }
+
+    fn get_internal_state(&self) -> ExecutionState {
+        self.get_internal_state()
+    }
+}
```

Comment:
Fair, oxade also mentioned something of the sort. I'll do some refactoring

Reply:
Because of the way paranoid_type_checks, gas_charging, etc. are baked into the execution code today, it's not immediately clear how we can abstract the error handling away from plugins and into interpreter. I think for now, plugins can construct the `VMError` and leverage the interpreter to position the error as they desire, and the interpreter will just check `plugin.is_critical` for whether to consume or raise the error. For example, in execute_main, ExitCode::Call branch, paranoid_type_checks' check_friend_or_private constructs an error slightly differently from gas_meter.charge_call. ignoring func.is_native() portion, in the current interpreter,
 paranoid_type_checks are baked into frame.make_call_frame, which get caught by the .map_errs.

Once the refactoring work is complete + I can guarantee that I didn't break anything, we can then go back in and lift the error handling out of the individual plugins (or, perhaps its fine for plugins to manage where and how they want to report errors)